run.go

```diff
@@ -13,48 +15,223 @@ import (
+
+				// we have to hold onto any panics here to be able to run AfterTest
+				var beforeTestErr any
+				if pkgTests.BeforeTest != nil {
+					func() {
+						defer func() {
+							if beforeTestErr = recover(); beforeTestErr != nil {
+								beforeTestErr = fmt.Sprintf("before test: %v\n\n%s", beforeTestErr, debug.Stack())
+							}
+						}()
+						pkgTests.BeforeTest(testHelperT)
+					}()
+				}
+
+				// only run the tests if any BeforeTest didn't panic
+				if beforeTestErr == nil {
+					res := make(chan TestResult)
+
+					wg := sync.WaitGroup{}
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+						for r := range res {
+							results[pkg][r.Name] = r
+						}
+					}()
+
+					runTest(pkg, test.Name, test.tester, res)
+					close(res)
+					wg.Wait()
+				} else {
+					results[pkg][name] = TestResult{
+						Package: pkg,
+						Name:    name,
+						Passed:  false,
+						Dur:     0,
+						Msgs: append(testHelperT.msgs, Msg{
+							Msg:   fmt.Sprintf("%v", beforeTestErr),
+							Level: LevelError,
+						}),
+					}
+				}
+
+				if pkgTests.AfterTest != nil {
+					var afterTestErr any
+					func() {
+						defer func() {
+							if afterTestErr = recover(); afterTestErr != nil {
+								afterTestErr = fmt.Sprintf("after test: %v\n\n%s", afterTestErr, debug.Stack())
+							}
+						}()
+						pkgTests.AfterTest(testHelperT)
+					}()
+
+					if afterTestErr != nil {
+						// replace test results with new one marking it failed and with this panic message.
+						old := results[pkg][name]
+						results[pkg][name] = TestResult{
+							Package: old.Package,
+							Name:    old.Name,
+							Passed:  false,
+							Dur:     old.Dur,
+							Msgs: append(old.Msgs, append(testHelperT.msgs, Msg{
+								Msg:   fmt.Sprintf("%v", afterTestErr),
+								Level: LevelError,
+							})...),
+						}
+					}
+				}
+			} else {
+				results[pkg][name] = TestResult{
+					Package: pkg,
+					Name:    name,
+					Passed:  false,
+					Dur:     0,
+					Msgs: append(pkgHelperT.msgs, Msg{
+						Msg:   fmt.Sprintf("%v", beforePkgErr),
+						Level: LevelError,
+					}),
+				}
+			}
 
 			return true
 		})
+
+		var afterPkgErr any
+		if pkgTests.AfterPackage != nil {
+			func() {
+				defer func() {
+					if afterPkgErr = recover(); afterPkgErr != nil {
+						afterPkgErr = fmt.Sprintf("after package: %v\n\n%s", afterPkgErr, debug.Stack())
+					}
+				}()
+				pkgTests.AfterPackage(pkgHelperT)
+			}()
+		}
+
+		// only update test results if AfterPackage panicked and BeforePackage did not
+		if beforePkgErr == nil && afterPkgErr != nil {
+			newResults := make(orderedmap.OrderedMap[string, TestResult])
+			results[pkg].Iterate(func(test string, result TestResult) bool {
+				newResults[test] = TestResult{
+					Package: result.Package,
+					Name:    result.Name,
+					Passed:  false,
+					Dur:     result.Dur,
+					Msgs: append(result.Msgs, append(pkgHelperT.msgs, Msg{
+						Msg:   fmt.Sprintf("%v", beforePkgErr),
```

Comment:
```suggestion
						Msg:   fmt.Sprintf("%v", afterPkgErr),
```

but why are we not reporting `afterPkgErr` if `beforePkgErr` is not nil?

Reply:
can't panic two things, and if before failed any failures in after could very well be related.

Reply:
wait, this isn't the path that panics, it should have already been reported above... though do the tests reflect that? hm.

Reply:
ah, I don't have any test cases with multiple things failing, let me add one. but I think it should have them both

Reply:
ok, yeah, this was easy enough to do in Run mode